File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-35.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        char = set() 
        l = 0 
        res = 0 
        for r in range(len(s)): 
            cur = s[r] 
            while(cur in char): 
                deleted = s[l]
                char.remove(deleted) 
                l += 1  

            char.add(cur)
            res = max(res, r - l + 1) 

        return res

File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-35.js
class Solution {
    /**
     * @param {string} s
     * @param {number} k
     * @return {number}
     */
    characterReplacement(s, k) { 
        let c = {}  

        let l = 0  
        let max = 0
        for (let r = 0; r < s.length; r ++) {
            const cur = s[r]  

            if (c[cur] === undefined) {
                c[cur] = 1
            } else { 
                c[cur] ++
            } 

            while(((r - l + 1) - Math.max(...Object.values(c))) > k ) {
                let deleted = s[l] 
                c[deleted] --  
                l ++
            }

           max = Math.max(max, (r - l + 1))
        } 

        return max
    }
}
